staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go

```diff
@@ -0,0 +1,273 @@
+}
+
+func (d *DynamicKMSEncryptionConfigContent) watchEncryptionConfigFile(stopCh <-chan struct{}) error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("error creating fsnotify watcher: %v", err)
+	}
+	defer watcher.Close()
+
+	if err = watcher.Add(d.filePath); err != nil {
+		return fmt.Errorf("error adding watch for file %s: %v", d.filePath, err)
+	}
+
+	for {
+		select {
+		case event := <-watcher.Events:
+			if err := d.handleWatchEvent(event, watcher); err != nil {
+				return err
+			}
+		case err := <-watcher.Errors:
+			return fmt.Errorf("received fsnotify error: %v", err)
+		case <-stopCh:
+			return nil
+		}
+	}
+}
+
+func (d *DynamicKMSEncryptionConfigContent) handleWatchEvent(event fsnotify.Event, watcher *fsnotify.Watcher) error {
+	// This should be executed after restarting the watch (if applicable) to ensure no file event will be missing.
+	defer d.queue.Add(workqueueKey)
+
+	// return if file has not removed or renamed.
+	if event.Op&(fsnotify.Remove|fsnotify.Rename) == 0 {
+		return nil
+	}
+
+	if err := watcher.Remove(d.filePath); err != nil {
+		klog.V(2).InfoS("Failed to remove file watch, it may have been deleted", "file", d.filePath, "err", err)
+	}
+	if err := watcher.Add(d.filePath); err != nil {
+		return fmt.Errorf("error adding watch for file %s: %v", d.filePath, err)
+	}
+
+	return nil
+}
+
+// runWorker to process file content
+func (d *DynamicKMSEncryptionConfigContent) runWorker() {
+	for d.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem processes file content when there is a message in the queue.
+func (d *DynamicKMSEncryptionConfigContent) processNextWorkItem() bool {
+	// key here is dummy item in the queue to trigger file content processing.
+	key, quit := d.queue.Get()
+	if quit {
+		return false
+	}
+	defer d.queue.Done(key)
+
+	err := d.loadEncryptionConfig()
+	if err == nil {
+		d.queue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with : %v", key, err))
+
+	// add dummy item back to the queue to trigger file content processing.
+	d.queue.AddRateLimited(key)
+
+	return true
+}
+
+// loadEncryptionConfig processes the next set of content from the file.
+func (d *DynamicKMSEncryptionConfigContent) loadEncryptionConfig() error {
+	encryptionConfig, err := d.readEncryptionFileContent(d.filePath)
+	if err != nil {
+		return err
+	}
+	if len(encryptionConfig) == 0 {
+		return fmt.Errorf("encryption config file is empty %q", d.filePath)
+	}
+
+	// compute hash for the encryption config file content.
+	encryptionConfigHash := d.computeHash(encryptionConfig)
+
+	// check if encryptionConfig is different from the current. Do nothing if they are the same.
+	if !d.hasEncryptionConfigChanged(encryptionConfigHash) {
+		return nil
+	}
+
+	err = d.processEncryptionConfig()
+	if err != nil {
+		return err
+	}
+
+	// update local copy of recent config content once processing is successful.
+	d.encryptionConfig.Store(encryptionConfigHash)
```

Comment:
Hash needs to come from `encryptionconfig.LoadEncryptionConfig`

```suggestion
	actualEncryptionConfigHash, err := d.processEncryptionConfig()
	if err != nil {
		return err
	}

	// update local copy of recent config content once processing is successful.
	d.encryptionConfig.Store(actualEncryptionConfigHash)
```